app/lib/account/default_auth_provider.dart

```diff
@@ -86,6 +93,69 @@ class DefaultAuthProvider extends BaseAuthProvider {
     final openIdData = await openIdDataFetch();
     return await token.verifySignature(openIdData.jwks);
   }
+
+  String _getOauthSiteAudience() =>
+      _oauthSiteAudience ?? activeConfiguration.pubSiteAudience!;
+
+  Future<String?> _getOauthSiteAudienceSecret(String audience) async {
+    return _oauthSiteAudienceSecret ??
+        await secretBackend.getCachedValue('${SecretKey.oauthPrefix}$audience');
+  }
+
+  @override
+  Future<Uri> getOauthAuthenticationUrl({
+    required String state,
+    required String nonce,
+  }) async {
+    // Using https://developers.google.com/identity/protocols/oauth2/web-server#httprest_1
+    return Uri.parse('https://accounts.google.com/o/oauth2/v2/auth').replace(
+      queryParameters: {
+        'client_id': _getOauthSiteAudience(),
+        'redirect_uri': getOauthCallbackUrl(),
+        'response_type': 'code',
+        'scope': [
+          oauth2_v2.Oauth2Api.openidScope,
+          oauth2_v2.Oauth2Api.userinfoEmailScope,
+          oauth2_v2.Oauth2Api.userinfoProfileScope,
+        ].join(' '),
+        'state': state,
+        'nonce': nonce,
+      },
+    );
+  }
+
+  @override
+  Future<AuthResult?> tryAuthenticateOauthCode({
+    required String code,
+  }) async {
+    try {
+      final audience = _getOauthSiteAudience();
+      final secret = await _getOauthSiteAudienceSecret(audience);
+      final tokenUri = Uri.parse('https://oauth2.googleapis.com/token');
+      final formData = {
+        'code': code,
+        'client_id': audience,
+        'client_secret': secret,
+        'redirect_uri': getOauthCallbackUrl(),
+        'grant_type': 'authorization_code',
+      };
+      final rs = await http.post(tokenUri, body: formData);
+      if (rs.statusCode != 200) {
+        return null;
+      }
+      final body = json.decode(rs.body);
+      // TODO: also expose access_token for domain verification calls
+      final idToken = body['id_token'] as String;
+      final auth = await _tryAuthenticateJwt(idToken);
+      if (auth == null) {
+        return null;
+      }
```

Comment:
should we verify the `nonce` here? I think it's part of the OIDC flow. So probably we should have the `tryAuten...` method fail, if nonce is not matched.